File: backend/baselines.py
"""
Baselines de rendimiento esperado ("lo esperado") por recurso.
---------------------------------------------------------------
Convierte un numero suelto en un diagnostico: "esto es bueno o malo?". Es la
idea de cuello de botella de Veeam: se compara lo MEDIDO contra lo ESPERADO.

Tres recursos:
  - disco    -> por tipo de almacenamiento (NVMe / SATA SSD / SAS HDD / SATA HDD)
  - red      -> por velocidad de enlace (1 / 10 / 25 / 40 GbE)
  - computo  -> throughput de compresion de referencia (por ahora, valor fijo)

IMPORTANTE: los valores son APROXIMADOS (ballpark), para dar contexto, no specs
exactas. Ajustar al hardware real del cliente.
"""

# --- Disco (unidad por test: seq=MB/s, rand=IOPS) ---------------------------
DISK_BASELINES = {
    "nvme-ssd": {"label": "NVMe SSD", "seqread": 3000, "seqwrite": 2200, "randread": 350000, "randwrite": 250000},
    "sata-ssd": {"label": "SATA SSD", "seqread": 550, "seqwrite": 500, "randread": 90000, "randwrite": 70000},
    "sas-hdd":  {"label": "SAS 10k HDD (RAID)", "seqread": 250, "seqwrite": 220, "randread": 800, "randwrite": 500},
    "sata-hdd": {"label": "SATA 7.2k HDD", "seqread": 170, "seqwrite": 150, "randread": 180, "randwrite": 140},
}
DEFAULT_DISK = "sata-ssd"

# --- Red (throughput efectivo aprox, ~0.94 de la tasa de linea, en MB/s) -----
NET_BASELINES = {
    "1gbe":  {"label": "1 GbE", "mbps": 118},
    "10gbe": {"label": "10 GbE", "mbps": 1180},
    "25gbe": {"label": "25 GbE", "mbps": 2950},
    "40gbe": {"label": "40 GbE", "mbps": 4720},
}
DEFAULT_NET = "10gbe"

# --- Computo (MB/s de compresion de referencia) -----------------------------
COMPUTE_EXPECTED_MBPS = 550

# Umbrales de veredicto (ratio medido/esperado).
_OK = 0.85
_WARN = 0.60


def verdict(ratio: float) -> str:
    if ratio >= _OK:
        return "ok"
    if ratio >= _WARN:
        return "warn"
    return "low"


def disk_catalog() -> list:
    return [{"key": k, "label": v["label"]} for k, v in DISK_BASELINES.items()]


def net_catalog() -> list:
    return [{"key": k, "label": v["label"]} for k, v in NET_BASELINES.items()]


def annotate_disk(results: list, baseline_key: str) -> list:
    """Agrega a cada fila de disco: esperado, unidad, % y veredicto.
    seq -> compara MB/s ; rand -> compara IOPS."""
    b = DISK_BASELINES.get(baseline_key, DISK_BASELINES[DEFAULT_DISK])
    for r in results:
        is_seq = r["name"].startswith("seq")
        measured = r["bw_mbps"] if is_seq else r["iops"]
        expected = b.get(r["name"], 0)
        ratio = (measured / expected) if expected else 0
        r["resource"] = "disk"
        r["expected"] = expected
        r["expected_unit"] = "MB/s" if is_seq else "IOPS"
        r["measured_metric"] = round(measured, 1)
        r["pct_of_expected"] = round(ratio * 100)
        r["status"] = verdict(ratio)
    return results


def annotate_net(measured_mbps: float, net_key: str) -> dict:
    b = NET_BASELINES.get(net_key, NET_BASELINES[DEFAULT_NET])
    expected = b["mbps"]
    ratio = measured_mbps / expected if expected else 0
    return {
        "resource": "net", "label": "Throughput de red",
        "value": round(measured_mbps, 1), "unit": "MB/s",
        "expected": expected, "expected_label": b["label"],
        "pct_of_expected": round(ratio * 100), "status": verdict(ratio),
    }


def annotate_compute(measured_mbps: float) -> dict:
    expected = COMPUTE_EXPECTED_MBPS
    ratio = measured_mbps / expected if expected else 0
    return {
        "resource": "compute", "label": "Compresion (CPU)",
        "value": round(measured_mbps, 1), "unit": "MB/s",
        "expected": expected, "pct_of_expected": round(ratio * 100),
        "status": verdict(ratio),
    }

File: internal/benchmark/baselines.go
package benchmark

import "math"

// Baselines de rendimiento esperado ("lo esperado") por recurso. Convierten un
// numero medido en un diagnostico (ok/warn/low) comparando contra lo esperado,
// que es la idea de cuello de botella de Veeam. Valores APROXIMADOS (ballpark).

type diskBaseline struct {
	Label                                  string
	SeqRead, SeqWrite, RandRead, RandWrite float64
}

var diskBaselines = map[string]diskBaseline{
	"nvme-ssd": {"NVMe SSD", 3000, 2200, 350000, 250000},
	"sata-ssd": {"SATA SSD", 550, 500, 90000, 70000},
	"sas-hdd":  {"SAS 10k HDD (RAID)", 250, 220, 800, 500},
	"sata-hdd": {"SATA 7.2k HDD", 170, 150, 180, 140},
}

// orden estable para el catalogo (los mapas de Go iteran al azar).
var diskOrder = []string{"nvme-ssd", "sata-ssd", "sas-hdd", "sata-hdd"}

const defaultDisk = "sata-ssd"

type netBaseline struct {
	Label string
	Mbps  float64
}

var netBaselines = map[string]netBaseline{
	"1gbe":  {"1 GbE", 118},
	"10gbe": {"10 GbE", 1180},
	"25gbe": {"25 GbE", 2950},
	"40gbe": {"40 GbE", 4720},
}

var netOrder = []string{"1gbe", "10gbe", "25gbe", "40gbe"}

const (
	defaultNet = "10gbe"
	okRatio    = 0.85
	warnRatio  = 0.60
)

// Exportados para los catalogos del handler (valores por defecto del selector).
const (
	DefaultDisk = defaultDisk
	DefaultNet  = defaultNet
)

type catalogEntry struct {
	Key   string `json:"key"`
	Label string `json:"label"`
}

func DiskCatalog() []catalogEntry {
	out := make([]catalogEntry, 0, len(diskOrder))
	for _, k := range diskOrder {
		out = append(out, catalogEntry{k, diskBaselines[k].Label})
	}
	return out
}

func NetCatalog() []catalogEntry {
	out := make([]catalogEntry, 0, len(netOrder))
	for _, k := range netOrder {
		out = append(out, catalogEntry{k, netBaselines[k].Label})
	}
	return out
}

func verdict(ratio float64) string {
	switch {
	case ratio >= okRatio:
		return "ok"
	case ratio >= warnRatio:
		return "warn"
	default:
		return "low"
	}
}

func expectedForTest(b diskBaseline, name string) float64 {
	switch name {
	case "seqread":
		return b.SeqRead
	case "seqwrite":
		return b.SeqWrite
	case "randread":
		return b.RandRead
	case "randwrite":
		return b.RandWrite
	}
	return 0
}

// annotateDisk agrega a cada fila: esperado, unidad, % y veredicto.
// seq -> compara MB/s ; rand -> compara IOPS.
func annotateDisk(rows []DiskRow, baselineKey string) []DiskRow {
	b, ok := diskBaselines[baselineKey]
	if !ok {
		b = diskBaselines[defaultDisk]
	}
	for i := range rows {
		r := &rows[i]
		isSeq := len(r.Name) >= 3 && r.Name[:3] == "seq"
		measured := r.Iops
		unit := "IOPS"
		if isSeq {
			measured = r.BwMbps
			unit = "MB/s"
		}
		expected := expectedForTest(b, r.Name)
		ratio := 0.0
		if expected > 0 {
			ratio = measured / expected
		}
		r.Resource = "disk"
		r.Expected = expected
		r.ExpectedUnit = unit
		r.MeasuredMetric = round1(measured)
		r.PctOfExpected = int(math.Round(ratio * 100))
		r.Status = verdict(ratio)
	}
	return rows
}
